File: 115-DistinctSubsequences/115-DistinctSubsequences.java
/*

115. Distinct Subsequences
Hard

Given a string S and a string T, count the number of distinct subsequences of S which equals T.

A subsequence of a string is a new string which is formed from the original string by deleting some (can be none) of the characters without disturbing the relative positions of the remaining characters. (ie, "ACE" is a subsequence of "ABCDE" while "AEC" is not).

It's guaranteed the answer fits on a 32-bit signed integer.

Example 1:

Input: S = "rabbbit", T = "rabbit"
Output: 3
Explanation:
As shown below, there are 3 ways you can generate "rabbit" from S.
(The caret symbol ^ means the chosen letters)

rabbbit
^^^^ ^^
rabbbit
^^ ^^^^
rabbbit
^^^ ^^^

Example 2:

Input: S = "babgbag", T = "bag"
Output: 5
Explanation:
As shown below, there are 5 ways you can generate "bag" from S.
(The caret symbol ^ means the chosen letters)

babgbag
^^ ^
babgbag
^^    ^
babgbag
^    ^^
babgbag
  ^  ^^
babgbag
    ^^^

*/

class Solution {
    public int numDistinct(String s, String t) {
        Integer[][] cache = new Integer[s.length() + 1][t.length() + 1];
        return dfs(cache, s, t, 0, 0);
    }
    private int dfs(Integer[][] cache, String s, String t, int si, int ti){
        if(cache[si][ti] != null) return cache[si][ti];
        if(ti == t.length()) return 1;
        if(si == s.length() && ti < t.length()) return 0;
        int ways = 0;
        if(s.charAt(si) == t.charAt(ti))
            ways += dfs(cache, s, t, si + 1, ti + 1);
        ways += dfs(cache, s, t, si + 1, ti);
        cache[si][ti] = ways;
        return cache[si][ti];
    }
}

File: 115-DistinctSubsequences/115-DistinctSubsequences.py
"""

115. Distinct Subsequences

Given two strings s and t, return the number of distinct subsequences of s which equals t.

The test cases are generated so that the answer fits on a 32-bit signed integer.

Example 1:

Input: s = "rabbbit", t = "rabbit"
Output: 3
Explanation:
As shown below, there are 3 ways you can generate "rabbit" from s.
rabbbit
rabbbit
rabbbit

Example 2:

Input: s = "babgbag", t = "bag"
Output: 5
Explanation:
As shown below, there are 5 ways you can generate "bag" from s.
babgbag
babgbag
babgbag
babgbag
babgbag

Constraints:

    1 <= s.length, t.length <= 1000
    s and t consist of English letters.

"""

class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        def dfs(si, ti):
            if ti == len(t): return 1
            if si == len(s): return 0
            if (si, ti) in cache: return cache[(si, ti)]
            res = 0
            if s[si] == t[ti]: res += dfs(si + 1, ti + 1)
            res += dfs(si + 1, ti)
            cache[(si, ti)] = res
            return res
        cache = {}
        return dfs(0, 0)
